Feature/letstryagain

kgiszewski opened the pull request.

kgiszewski commented:
![image](https://user-images.githubusercontent.com/4653907/30434103-56a2193c-9934-11e7-8b97-4b951c8d375b.png)

![image](https://user-images.githubusercontent.com/4653907/30434110-59c6e8a4-9934-11e7-96cb-c091f72e4545.png)

![image](https://user-images.githubusercontent.com/4653907/30434117-5cc5dace-9934-11e7-9ca8-c6c2e8d4e920.png)

![image](https://user-images.githubusercontent.com/4653907/30434122-5f1d4550-9934-11e7-90e6-80ef51eddc8c.png)

![image](https://user-images.githubusercontent.com/4653907/30434188-98ddf712-9934-11e7-8e02-1473a7e80525.png)


```var UrlPickerTemplate = {};

UrlPickerTemplate.getTitle = function (value, scope) {
    //this is the property model
    if (value.length) {
        var firstValue = value[0];
        return ArchetypeSampleLabelTemplates.UrlPicker(firstValue, scope, { propertyName: "name" });
    }

    //if you wanted to get the name of the content instead, you'd have to get it from the server here since it's not in the model

    return "";
};
```

All of these scenarios are working.

kgiszewski commented:
@Nicholas-Westby I promised all the things. I think it's much much much cleaner now that my first 15 attempts :)

I do wanna know if there's a better syntax for `var deferred = $q.defer();` Seems boilerplate but maybe necessary.

kgiszewski commented:
oh and I refactored a lot of the cache to work on promises as my original code seemed to make several blunders.

kgiszewski commented:
I've been testing the crap out of this, I think I'd like to release this but would like @Nicholas-Westby to maybe review this weekend :)

Nicholas-Westby commented:
Just got home from an out of town trip, but hopefully I can take a look at this tomorrow night.

kgiszewski commented:
@Nicholas-Westby Thanks dood. I'm looking for validation in my solution or complete and utter ridicule if I'm way off. All seems to be working, however this could mean a breaking change for some folks. I think the approach of DI is ace and will nudge people in that direction. I've rewritten all the built-ins to be promised based and changed the samples in code to do the same.

Nicholas-Westby commented on app/helpers/sampleLabelHelpers.js:
```diff
@@ -1,70 +1,47 @@
+    return function ($q, archetypeCacheService) {    
+        var deferred = $q.defer();
+    
+        archetypeCacheService.getEntityById(1054, 'document').then(function(entity) {
```
Should "1054" be hardcoded here? Unsure of what "testEntityPromise2" is supposed to do, so can't be sure if this is correct or not.

kgiszewski replied:
Ill make it nicer. The intent was to show the cache service injected. 👍

Nicholas-Westby commented:
@kgiszewski I only had an hour or so, but I gave it a very quick review and test run. Good stuff. 👍 

I made a few recommended changes here (nothing all that necessary, but might be good to include): https://github.com/Nicholas-Westby/Archetype/commits/feature/letstryagain

Nicholas-Westby commented on app/services/archetypeCacheService.js:
```diff
@@ -1,106 +1,82 @@
+        getEntityById: function(id, type) {
+            var deferred = $q.defer();
+            
+            var cachedEntity = _.find(entityCache, function (e){
```
Why not make `entityCache` an associative array (with the ID as key) rather than a plain array? Could save some CPU cycles by performing instant lookup rather than scanning every item in the array for a match.

kgiszewski replied:
Great feedback, sounds like a quick win. 👍

kgiszewski commented:
@Nicholas-Westby I have updated the PR with your suggestions: https://github.com/kgiszewski/Archetype/pull/431/commits/46b0396652289d663e3f30daece6341821a34599
